tests/trainer/test_checkpoint.py

```diff
@@ -109,6 +138,100 @@ def inject_stateful_callback_hparams(monkeypatch: MonkeyPatch):
+    - assert that the model weights are the original model, even though the optimizer and scheduler are different.
+    """
+    del world_size  # unused. Read via env variable
+    if not isinstance(mosaic_trainer_hparams.train_dataset, SyntheticHparamsMixin):
+        pytest.skip("Checkpointing tests require synthetic data")
+        return
+    if not isinstance(mosaic_trainer_hparams.val_dataset, SyntheticHparamsMixin):
+        pytest.skip("Checkpointing tests require synthetic data")
+        return
+    mosaic_trainer_hparams.device = device_hparams
+    mosaic_trainer_hparams.train_dataset.use_synthetic = True
+    mosaic_trainer_hparams.train_dataset.shuffle = False
+    mosaic_trainer_hparams.val_dataset.use_synthetic = True
+    mosaic_trainer_hparams.val_dataset.shuffle = False
+    mosaic_trainer_hparams.grad_accum = 2
+    mosaic_trainer_hparams.loggers = []
+    mosaic_trainer_hparams.train_batch_size = 8
+    mosaic_trainer_hparams.eval_batch_size = 16
+    mosaic_trainer_hparams.max_epochs = 2
+    mosaic_trainer_hparams.precision = Precision.FP32
+    mosaic_trainer_hparams.callbacks = [DummyStatefulCallbackHparams(), EventCounterCallbackHparams()]
+    mosaic_trainer_hparams.train_subset_num_batches = 5
+    mosaic_trainer_hparams.device = device_hparams
+    checkpoint_a_folder = "first"
+    mosaic_trainer_hparams.save_checkpoint = CheckpointSaverHparams(
+        interval_unit="ep" if checkpoint_filename.startswith("ep") else "it",
+        interval=1,
+        folder=checkpoint_a_folder,
+    )
+    mosaic_trainer_hparams.seed = seed
+    mosaic_trainer_hparams.validate_every_n_batches = validate_every_n_batches
+    mosaic_trainer_hparams.validate_every_n_epochs = validate_every_n_epochs
+    final_checkpoint = "ep2.pt" if checkpoint_filename.startswith("ep") else "it8.pt"
+    _test_checkpoint_trainer(mosaic_trainer_hparams)
+
+    trainer_1_hparams_filepath = run_directory.get_relative_to_run_directory(checkpoint_a_folder, "hparams.yaml")
+
+    # re-create the trainer from the YAML
+    second_trainer_hparams = TrainerHparams.create(trainer_1_hparams_filepath, cli_args=False)
+
+    checkpoint_a_file_path = run_directory.get_relative_to_run_directory(checkpoint_a_folder, final_checkpoint)
+
+    # load only model weights
+    second_trainer_hparams.load_checkpoint = CheckpointLoaderHparams(filepath=checkpoint_a_file_path,
+                                                                     load_weights_only=True,
+                                                                     strict_model_weights=True)
+    # setup a new optimizer
+    second_trainer_hparams.optimizer = AdamWHparams()
+
+    # setup a new LR scheduler
+    scheduler_options = [ConstantLRHparams(), CosineAnnealingLRHparams(T_max=f"{second_trainer_hparams.max_epochs}ep")]
+    second_trainer_hparams.schedulers = [random.choice(scheduler_options)]
```

Comment:
@moinnadeem just ran into this now - it's very strange to use randomness in a test in this way since it can potentially cause flaky tests. Was there a reason you wanted this randomness here?

Reply:
Ah, I see the rationale. I just wanted to make sure that we test several schedulers, but didn't think it was worth the time to test all of them. In hindsight, we should either pick the more difficult one, or do both. I agree with you.

Reply:
I'm modifying this line in a PR for a different purpose anyways, so I'll fix it here.

Reply:
Thanks Jamie!